mitodl/mitxonline
# 869 add unique key to ecommerce line and ecommerce transaction

#### What are the relevant tickets?
https://github.com/mitodl/mitxonline/issues/869

#### What's this PR do?

- Adding uniqueness to `ecommerce_line` and its related code changes to prevent adding duplicates
- Adding a new column `transaction_id` to `ecommerce_transaction`(transaction_type exists), populating it with existing data, and finally making it unique, so 3 migrations for this column change
- Updating the existing automated tests due to unique key changes


#### How should this be manually tested?

- Before running the migration, dedup any existing data that violates the uniqueness:

(I've already cleaned up the dups on the RC database, but I will double-check before merging to RC)
``` 
SELECT COUNT(*), order_id, purchased_content_type_id, purchased_object_id FROM ecommerce_line GROUP BY order_id,purchased_content_type_id, purchased_object_id HAVING COUNT(*) >1;

SELECT COUNT(*) FROM ecommerce_transaction WHERE data->>'transaction_id' is null AND data->>'id' is null;

```

- After fixing the existing data, then run the migrate
- transaction_id should be populated for all the _existing_ transaction http://mitxonline.odl.local:8013/admin/ecommerce/transaction/

- Go to frontend, and enroll in a course. It should go through.

<img width="1369" alt="image" src="https://user-images.githubusercontent.com/3138890/185953954-54ca5932-ac6e-4e73-b60b-301ca59e91b0.png">

- Check transaction admin (transaction Id should be populated for _new_ record)

<img width="911" alt="image" src="https://user-images.githubusercontent.com/3138890/185954879-bbe83acc-5c2e-472d-a96f-aa71a5d269e3.png">

- Go to Fulfilled orders from admin, and request a refund, you should see transaction_id is populated for _new_ refund as well

![image](https://user-images.githubusercontent.com/3138890/185955683-0ea84e9b-f255-47b9-8b13-b6e0ddc712a4.png)



#### Where should the reviewer start?
(Optional)

#### Any background context you want to provide?
(Optional)

#### Screenshots (if appropriate)
(Optional)
